File: 303_RangeSumQuery-Immutable.py
# 45.07%
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.numSums = [0] * (len(nums) + 1)
        for i, num in enumerate(nums):
            self.numSums[i + 1] = self.numSums[i] + num

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.numSums[j + 1] - self.numSums[i]


# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# param_1 = obj.sumRange(i,j)

# A solution from LeetCode, seems better than mine
class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.accu = [0]
        for num in nums: 
            self.accu += self.accu[-1] + num,

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int 
        :type j: int
        :rtype: int 
        """
        return self.accu[j + 1] - self.accu[i]

File: 303_RangeSumQuery-Immutable.cpp
//28.91%
class NumArray {
public:
    NumArray(vector<int> nums): sumNums(nums.size() + 1, 0) {
        for(int i = 0; i < nums.size(); i++){
			sumNums[i + 1] = sumNums[i] + nums[i];
		}
    }
    
    int sumRange(int i, int j) {
        return sumNums[j + 1] - sumNums[i];
    }
private:
	vector<int> sumNums;
};

/**
 * Your NumArray object will be instantiated and called as such:
 * NumArray obj = new NumArray(nums);
 * int param_1 = obj.sumRange(i,j);
 */
 
//Other solution from LeetCode
class NumArray {
public:
    NumArray(vector<int> &nums) {
        accu.push_back(0);
        for (int num : nums)
            accu.push_back(accu.back() + num);
    }

    int sumRange(int i, int j) {
        return accu[j + 1] - accu[i];
    }
private:
    vector<int> accu;
};

//Use standard algorithm 'partial_sum'
class NumArray {
public:
    NumArray(vector<int> &nums) : psum(nums.size()+1, 0) {
        partial_sum( nums.begin(), nums.end(), psum.begin()+1);
    }

    int sumRange(int i, int j) {
        return psum[j+1] - psum[i];
    }
private:
    vector<int> psum;    
};